src/osm4routing/reader.rs

```diff
@@ -1,259 +1,434 @@
-    assert_eq!(2, r.nodes[&NodeId(1)].uses);
-    assert_eq!(1, r.nodes[&NodeId(2)].uses);
-    assert_eq!(2, r.nodes[&NodeId(3)].uses);
 
-    assert_eq!(2, r.nodes().len());
-}
+    self.reader.ways.push(way);
 
-#[test]
-fn test_split() {
-    let mut nodes = HashMap::new();
-    nodes.insert(NodeId(1), Node::default());
-    nodes.insert(NodeId(2), Node::default());
-    nodes.insert(NodeId(3), Node::default());
-    nodes.insert(NodeId(4), Node::default());
-    nodes.insert(NodeId(5), Node::default());
-    let ways = vec![
-        Way {
-            id: WayId(0),
-            nodes: vec![NodeId(1), NodeId(2), NodeId(3)],
-            properties: EdgeProperties::default(),
-        },
-        Way {
-            id: WayId(0),
-            nodes: vec![NodeId(4), NodeId(5), NodeId(2)],
-            properties: EdgeProperties::default(),
-        },
-    ];
-    let mut r = Reader {
-        nodes,
-        ways,
-        ..Default::default()
+    Ok(())
+  }
+
+  /// Consumes a node from the given CSV record.
+  /// 
+  /// # Arguments
+  /// 
+  /// * `record` - The CSV record containing the node data in string format.
+  /// 
+  /// # Returns
+  /// 
+  /// An empty `Result` indicating success or an error as `Box<dyn Error>`.
+  fn consume_node(&mut self, record: &StringRecord) -> Result<(), Box<dyn Error>> {
+    let node_id: NodeId = NodeId(record[0].parse::<i64>()?);
+    let lat: f64 = record[1].parse::<f64>()?;
+    let lon: f64 = record[2].parse::<f64>()?;
+
+    let node = Node {
+      id: node_id,
+      coord: Coord { lat, lon },
+      uses: 0,
     };
-    r.count_nodes_uses();
-    let edges = r.edges();
-    assert_eq!(3, edges.len());
+
+    self.reader.nodes.insert(node_id, node);
+
+    Ok(())
+  }
 }
 
-#[test]
-fn test_wrong_file() {
-    let r = read("i hope you have no file name like this one");
-    assert!(r.is_err());
+pub struct OsmReader {
+  reader: Reader,
 }
 
-#[test]
-fn forbidden_values() {
-    let (_, ways) = Reader::new()
-        .reject("highway", "secondary")
-        .read("src/osm4routing/test_data/minimal.osm.pbf")
-        .unwrap();
-    assert_eq!(0, ways.len());
+impl OsmReader {
+  pub fn new() -> OsmReader {
+    OsmReader {
+      reader: Reader::new(),
+    }
+  }
+
+  pub fn reject(mut self, key: &str, value: &str) -> Self {
+    self.reader = self.reader.reject(key, value);
+    self
+  }
+
+  /// Reads the nodes and ways from the given OSM files and returns them.
+  /// 
+  /// # Arguments
+  /// 
+  /// * `filename` - The path to the OSM file containing node data.
+  /// 
+  /// # Returns
+  /// 
+  /// A tuple containing vectors of nodes and edges, wrapped in a `Result`. If successful,
+  /// the vectors are returned. Otherwise, an error is returned as `Box<dyn Error>`.
+  fn read(&mut self, filename: &str) -> Result<(Vec<Node>, Vec<Edge>), Box<dyn Error>> {
+    let path = std::path::Path::new(filename);
+    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
```

Comment:
Why the Error -> String -> Error conversion?

Consider using the anyhow library.

Reply:
Actually, this is fine. Anyhow provides a conversion if you'd like. Something like:

```
fn foo() -> anyhow::Result<File> {
    return File::open("/foo/bar").map_err(|e| e.into());
}
```